engines/grim/hotspot.cpp

```diff
@@ -0,0 +1,768 @@
+	for (int i = 0; i < 8; ++i) {
+		_flashBitmaps[i] = Bitmap::create(Common::String::format("cursor%d_hl.tga", i));
+		_flashBitmaps[i]->_data->load();
+		_flashBitmaps[i]->_data->_hasTransparency = true;
+	}
+}
+
+int HotspotMan::addHotspot(const Common::String &name, const Math::Vector3d &pos, const Common::String &scene) {
+	HotObject hs;
+	hs._pos = pos;
+	hs._active = false;
+	_hotobject.push_back(hs);
+
+	int num = _hotobject.size() - 1;
+	renameHotspot(num, name);
+	return num;
+}
+
+void HotspotMan::disableAll() {
+	for (size_t i = 0, total = _hotobject.size(); i < total; ++i)
+		_hotobject[i]._active = false;
+}
+
+void HotspotMan::flashHotspots() {
+	if (_cutScene > 0)
+		return;
+	_flashStart = g_system->getMillis();
+	_flashHS = true;
+	update();
+}
+
+void HotspotMan::drawActive() {
+	if (_flashHS) {
+		unsigned int curTime = g_system->getMillis();
+		unsigned int delta = curTime - _flashStart;
+		if (delta >= 1700) _flashHS = false;
+		if ((delta % 900) < 450) {
+			int setup = g_grim->getCurrSet()->getSetup();
+			Common::Array<Hotspot> &hotspots = _hotspots[activeSet()];
+			for (size_t i = 0, total = hotspots.size(); i < total; ++i) {
+				Hotspot &hs = hotspots[i];
+				if (hs._setup == setup && (hs._objId<=0 || _hotobject[hs._objId]._active)) {
+					if ((_ctrlMode == Dialog && hs._type >= 20) ||
+						(_ctrlMode == Special && hs._type >= 10 && hs._type < 20) ||
+						(_ctrlMode == Linear && hs._type >= 10 && hs._type < 20) ||
+						((_ctrlMode == Normal || _ctrlMode == NoWalk) && hs._type < 10)) {
+						int cu = hs._type % 10;
+						if (hs._type == 3) cu = 2;
+						if (hs._type == 4) cu = 1;
+						Common::Point p = hs._region.center();
+						_flashBitmaps[cu]->draw(p.x - 15, p.y - 15);
+					}
+				}
+			}
+		}
+	}
+
+	if (_ctrlMode == Inventory) {
+		const int dx = 4;
+		const int numRows = (_inventory.size() - 1) / _cols + 1;
+		int x1 = _x0 + _w * _cols, y1 = _y0 + _h * numRows;
+		g_driver->blackbox(0, 0, 640, 480, 0.4);
+		g_driver->blackbox(_x0 - dx, _y0 - dx, x1 + dx, y1 + dx, 0.9);
+		PrimitiveObject rect;
+		Color col(195, 195, 100);
+		rect.createRectangle(Common::Point(_x0 - dx, _y0 - dx), Common::Point(x1 + dx, y1 + dx), col, false);
+		rect.draw();
+
+		for (int j = 0, idx = 0; j < _rows; j++) {
+			for (int i = 0; i < _cols; ++i, ++idx) {
+				if (idx >= (int)_inventory.size())
+					break;
+				_inventory[idx]._bmp->draw(_x0 + i * _w, _y0 + j * _h);
+			}
+		}
+	}
+}
+
+void HotspotMan::updatePerspective() {
+	g_grim->getCurrSet()->setupCamera();
+	for (size_t i = 0; i < _hotobject.size(); ++i) {
+		int x = 0, y = 0;
+		g_driver->worldToScreen(_hotobject[i]._pos, x, y);
+		_hotobject[i]._rect = Common::Rect(x - 10, y - 10, x + 10, y + 10);
+	}
+	restoreCursor();
+}
+
+Common::String HotspotMan::activeSet() {
+	return (_curOption.empty() || _ctrlMode != Options) ?
+				g_grim->getCurrSet()->getName() : _curOption;
+}
+
+// get min distance s for x0 + s * x1, y0 + t * y1
+static double lineLineDist(
```

Comment:
I think we have something in math already

Reply:
Where?

Reply:
never mind